presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -311,47 +406,160 @@ private void processPage(Page page)
 
         public boolean isFull()
         {
-            long memorySize = groupByHash.getEstimatedSize();
-            for (Aggregator aggregator : aggregators) {
-                memorySize += aggregator.getEstimatedSize();
-            }
-            memorySize -= operatorContext.getOperatorPreAllocatedMemory().toBytes();
-            if (memorySize < 0) {
-                memorySize = 0;
+            if (isBusy()) {
+                return true;
             }
+
+            long memorySize = getSizeInMemory();
             if (partial) {
                 return !operatorContext.trySetMemoryReservation(memorySize);
             }
             else {
-                operatorContext.setMemoryReservation(memorySize);
+                if (isSpillingEnabled()) {
+                    systemMemoryContext.setBytes(memorySize);
+
+                    if (shouldSpill(memorySize)) {
+                        log.info(
+                                "Started spilling memory %s",
+                                DataSize.succinctBytes(memorySize));
+
+                        spillToDisk();
```

Comment:
This method was somehow wrong previously. Memory accounting doesn't fit in this method from the same reasons.

Renamed to `checkFullAndSetMemory`